src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -161,43 +164,49 @@ internal Span<byte> GetSpan(int sizeHint)
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.minimumSize);
             }
 
-            lock (_sync)
+            // If writing is currently active and enough space, don't need to take the lock to just set WritingActive.
+            // IsWritingActive is needed to prevent the reader releasing the writers memory when it fully consumes currently written.
+            if (!_operationState.IsWritingActive ||
+                _writingMemory.Length == 0 || _writingMemory.Length < sizeHint)
```

Comment:
I think I like this. How is the perf if you move the if condition inside WriteHeadSynchronized? Obviously you still wouldn't take the lock unless necessary.

Reply:
Would introduce an extra "always" call?

Could do an extra shared method like:
```
AllocateWriteHeadIfNeeded
```
Then force inline it?

Reply:
https://github.com/dotnet/corefx/pull/35234/commits/55e365012d84a9f6ed37bfebc74a694bee84d120